### PR TITLE
feat(matches): add overview page

### PR DESCRIPTION
**Pull request by LarsSelbekk**
- chore: bump bl-model
- refactor(api): improve type safety of api SWR fetchers
- build(tsconfig): make unused things not error
- refactor(theme): add typography variants
- a11y: add html lang attribute
- feat(matches): add matches overview page
- feat(matches): poll matches for changes every 5s
- hack(match): REVERT ME LATER fix type errors

### CODE REVIEW

**Comment by LarsSelbekk** [comment]
noMatches

oneMatchNotDone
partialMatch
combo
big
moreCombo
moreGroups
allDone

**Comment by LarsSelbekk** [comment]
You should probably read the commit messsages here

File: .eslintrc.js [hunk 1]
```diff
@@ -53,5 +53,6 @@ module.exports = {
         },
       },
     ],
+    "unicorn/no-null": "off",
```
Comment (AdrianAndersen) [inline .eslintrc.js:56]: why?

File: src/components/match/matchesList/MatchesList.tsx [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    apiFetcher<MatchWithDetails[]>
+  );
+
+  if (tokenError || matchesError) {
+    return <Box>En feil har oppstått.</Box>;
+  }
+
+  if (matches === undefined) {
+    return <Box>Laster</Box>;
+  }
```
Comment (AdrianAndersen) [inline src/components/match/matchesList/MatchesList.tsx:33]: You should use skeletons here. Really quick to add
```javascript
<Skeleton />
```

File: src/components/match/matchesList/MatchesList.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  const userMatches = matches
+    .filter((match) => match._variant === MatchVariant.UserMatch)
+    .map((userMatch) => userMatch as UserMatchWithDetails)
+    .sort((a, b) => (matchFulfilled(a) ? 1 : matchFulfilled(b) ? -1 : 0));
+  const standMatchesByTime = groupMatchesByTimeAndLocation(standMatches);
+  const userMatchesByTime = groupMatchesByTimeAndLocation(userMatches);
+
+  if (matches.length === 0) {
+    return <Box>Du har ingen overleveringer :)</Box>;
+  }
```
Comment (AdrianAndersen) [inline src/components/match/matchesList/MatchesList.tsx:48]:
```Javascript
<Alert severity="info">Du har ingen overleveringer :)</Alert>
```

File: src/components/match/matchesList/helper.tsx [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  if (hasPickupItems) {
+    s.push("motta ");
+    if (pickupItems === 1) {
+      s.push("én bok");
+    } else {
+      s.push(`${pickupItems} bøker`);
+    }
+  }
+  return s.join("");
+}
```
Comment (AdrianAndersen) [inline src/components/match/matchesList/helper.tsx:56]: Seems like there is a lot of duplication in this method, why not just:
```javascript
export function formatActionsString(handoffItems: number, pickupItems: number) {
  const hasHandoffItems = handoffItems > 0;
  const hasPickupItems = pickupItems > 0;
  const actions: string[] = [];

  if (hasHandoffItems) {
    const handoffItemText = handoffItems === 1 ? "én bok" : `${handoffItems} bøker`;
    actions.push(`Du skal levere ${handoffItemText}`);
    if (hasPickupItems) {
      actions.push("og");
    }
  }

  if (hasPickupItems) {
    const pickupItemText = pickupItems === 1 ? "én bok" : `${pickupItems} bøker`;
    actions.push(`motta ${pickupItemText}`);
  }

  return actions.join(" ");
}
```

Not properly tested this btw, so please check that it works as before

File: src/components/match/matchesList/helper.tsx [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+export function groupMatchesByTimeAndLocation<T extends MatchWithDetails>(
+  matches: T[]
```
Comment (AdrianAndersen) [inline src/components/match/matchesList/helper.tsx:59]: no docs?

File: src/components/match/matchesList/helper.tsx [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  return keys;
+}
+
+export function formatDatetime(date: Date): string {
+  return (
+    date.toLocaleDateString("no") +
+    " " +
+    date.toLocaleTimeString("no", { timeStyle: "short" })
+  );
+}
```
Comment (AdrianAndersen) [inline src/components/match/matchesList/helper.tsx:110]: Would prefer template literals here

File: src/pages/match/index.tsx [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import { Typography } from "@mui/material";
+
+const MatchesPage: NextPage = () => {
+  return (
+    <>
+      <Head>
+        <title>Mine overleveringer | Boklisten.no</title>
+        <meta name="description" content="Overleveringer av bøker" />
+      </Head>
+      <div style={{ padding: "1em" }}>
```
Comment (AdrianAndersen) [inline src/pages/match/index.tsx:14]: why not rem?

File: src/utils/types.ts [hunk 8]
```diff
@@ -47,3 +49,24 @@ export interface Item {
...
+interface RelevantMatchedPersonDetails {
+  name: string;
+}
+
+export type MatchWithDetails =
+  | StandMatch
+  | (UserMatch & {
+      senderDetails: RelevantMatchedPersonDetails;
+      receiverDetails: RelevantMatchedPersonDetails;
+    });
```
Comment (AdrianAndersen) [inline src/utils/types.ts:62]: these are used by both bl-api and bl-next, and should therefore be appended to bl-model

**Review by AdrianAndersen — CHANGES_REQUESTED** [review]
Looks good for now, even though we might iterate on the design later. I was kinda picky about some things, but they should be really quick to implement/fix.

File: src/components/match/matchesList/MatchesList.tsx [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+import { apiFetcher } from "api/api";
+import { getAccessTokenBody } from "../../../api/token";
+import { groupMatchesByTimeAndLocation, matchFulfilled } from "./helper";
+import { MatchVariant, StandMatch } from "@boklisten/bl-model";
+import { MatchListItemGroups } from "./MatchListItemGroups";
+import ProgressBar from "./ProgressBar";
+import { Box } from "@mui/material";
+import { MatchWithDetails, UserMatchWithDetails } from "../../../utils/types";
+
+// TODO: real-time updating when receiver is scanning books?
```
[the remark this answers is not in the document]
Comment (LarsSelbekk) [reply src/components/match/matchesList/MatchesList.tsx]: Wasn't that removed in the latest commit? Oops if not

File: src/pages/match/index.tsx (hunk 7 above)
Comment (LarsSelbekk) [reply src/pages/match/index.tsx:14]: Oopsie

File: .eslintrc.js (hunk 1 above)
Comment (LarsSelbekk) [reply .eslintrc.js:56]: Firstly I strongly disagree with the premise of the rule that undefined can be used instead, and secondly it makes it really hard to interface with bl-model types which use null

Comment (AdrianAndersen) [reply .eslintrc.js:56]: Sure, I kinda get that 😛 Can disable it for now at least

File: src/components/match/matchesList/MatchesList.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (AdrianAndersen) [reply src/components/match/matchesList/MatchesList.tsx]: ah yeah. I went commit by commit, so did not see that it was removed when I encountered this

File: src/components/match/matchesList/MatchesList.tsx (hunk 3 above)
Comment (AdrianAndersen) [reply src/components/match/matchesList/MatchesList.tsx:48]: <Alert/> is not a popup, just a styled Box:
image

File: src/components/match/matchesList/helper.tsx (hunk 4 above)
Comment (AdrianAndersen) [reply src/components/match/matchesList/helper.tsx:56]: Yeah, I prefer the concatted version. So up to you, but at the very least you should use a proper variable name

File: src/components/match/matchesList/helper.tsx (hunk 6 above)
Comment (AdrianAndersen) [reply src/components/match/matchesList/helper.tsx:110]: It becomes clean if you extract the logic into variables, and then use template literals

**Comment by LarsSelbekk** [comment]
noMatches

oneMatchNotDone
partialMatch
combo
big
moreCombo
moreGroups
allDone

File: src/components/match/MatchOverview.tsx [hunk 10]
```diff
@@ -12,7 +12,7 @@ import {
 } from "@mui/material";
```
[the remark this answers is not in the document]
Comment (LarsSelbekk) [reply src/components/match/MatchOverview.tsx:12]: Done

File: src/components/match/matchesList/MatchesList.tsx [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+  );
+  const userId = accessToken?.details;
+  const { data: matches, error: matchesError } = useSWR(
+    `${BL_CONFIG.collection.match}/me`,
+    apiFetcher<MatchWithDetails[]>
+  );
+
+  if (tokenError || matchesError) {
+    return <Box>En feil har oppstått.</Box>;
+  }
```
[the remark this answers is not in the document]
Comment (LarsSelbekk) [reply src/components/match/matchesList/MatchesList.tsx]: Done in cb6b63d

File: src/components/match/matchesList/MatchesList.tsx (hunk 2 above)
Comment (LarsSelbekk) [reply src/components/match/matchesList/MatchesList.tsx]: Done in cb6b63d

File: src/pages/match/index.tsx (hunk 7 above)
Comment (LarsSelbekk) [reply src/pages/match/index.tsx]: Done in cb6b63d

File: src/components/match/matchesList/helper.tsx (hunk 6 above)
Comment (LarsSelbekk) [reply src/components/match/matchesList/helper.tsx]: Done in cb6b63d

File: src/components/match/matchesList/helper.tsx (hunk 5 above)
Comment (LarsSelbekk) [reply src/components/match/matchesList/helper.tsx:67]: Done in cb6b63d

File: src/components/match/matchesList/helper.tsx (hunk 4 above)
Comment (LarsSelbekk) [reply src/components/match/matchesList/helper.tsx]: Name changed in cb6b63d

File: src/components/match/matchesList/MatchesList.tsx (hunk 3 above)
Comment (LarsSelbekk) [reply src/components/match/matchesList/MatchesList.tsx]: Done in cb6b63d

**Review by AdrianAndersen — APPROVED** [review]

**Comment by LarsSelbekk** [comment]
Wouldn't it be better to have a single arrival endpoint from bl-web which can handle the token catching?

**Comment by AdrianAndersen** [comment]
["Mine Overleveringer"-side](https://trello.com/c/C7twbaZN/624-mine-overleveringer-side)

**Comment by AdrianAndersen** [comment]
> Wouldn't it be better to have a single arrival endpoint from bl-web which can handle the token catching?

Yeah, I thought about that too. I have some plans for slowly integrating bl-next into bl-web, and in that case we really should do that yeah. I think the solution in this PR is ok for now, but definitely something we will look into this fall! 🚀

**Pull request merged**